### PR TITLE
feat(focus-mvp-android-device): Add Focus Visualization classes

### PR DESCRIPTION
**Pull request by brocktaylor7**
#### Details

This creates the FocusVisualizer, FocusVisualizerStyles, FocusCanvasView, FocusElementHighlight, and FocusElementLine classes to handle the full lifecycle of tabstops visualizations for the tabstops test.

##### Motivation

to be able to draw visualizations for the Focus Order MVP.

#### Pull request checklist

- [n/a] Addresses an existing issue: #0000
- [x] Added/updated relevant unit test(s)
- [x] Ran `./gradlew fastpass` from `AccessibilityInsightsForAndroidService`
- [x] PR title _AND_ final merge commit title both start with a semantic tag (`fix:`, `chore:`, `feat(feature-name):`, `refactor:`).

### CODE REVIEW

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import android.graphics.Rect;
+import android.view.View;
+import android.view.accessibility.AccessibilityNodeInfo;
+import java.util.HashMap;
+
+public class FocusElementHighlight extends View {
+  private static final String TAG = "FocusElementHighlight";
+  public AccessibilityNodeInfo eventSource;
+  private int yOffset;
+  public int tabStopCount;
```
Comment (waabid) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java:18]: These variables could likely all be private as far as I can tell.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    this.drawInnerCircle(
+        xCoordinate, yCoordinate, radius, currentPaints.get("innerCircle"), canvas);
+    this.drawNumberInCircle(
+        xCoordinate, yCoordinate, tabStopCount, currentPaints.get("number"), canvas);
+    this.drawOuterCircle(
+        xCoordinate, yCoordinate, radius, currentPaints.get("outerCircle"), canvas);
+  }
+
+  public int getYOffset() {
```
Comment (waabid) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java:105]: This method and the ones below that are only called in this class should likely be private.

Same also applies to FocusElementLine.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+  public FocusElementHighlight(
+      AccessibilityNodeInfo eventSource,
+      HashMap<String, Paint> currentPaints,
+      int radius,
+      int tabStopCount,
+      View view) {
+    this.view = view;
+    this.eventSource = eventSource;
+    this.yOffset = setYOffset();
```
Comment (DaveTryon) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java:33]: Would it make sense to call `updateWithNewCoordinates` here, instead of having separate calls to `setYOffset` and `setCoordinates` ? That gives us a single place where we handle any position-related stuff

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+import android.graphics.Canvas;
+import android.graphics.Paint;
+import android.graphics.Rect;
+import android.view.View;
+import android.view.accessibility.AccessibilityNodeInfo;
+import java.util.HashMap;
+
+public class FocusElementHighlight {
+  private static final String TAG = "FocusElementHighlight";
```
Comment (DaveTryon) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java:14]: IIRC, we only use `TAG` when we're logging. If no logging is being added, can it go away? (applies to 3 classes)

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import android.view.View;
+import android.view.accessibility.AccessibilityNodeInfo;
+import java.util.HashMap;
+
+public class FocusElementLine {
+  private static final String TAG = "FocusElementLine";
+  private AccessibilityNodeInfo eventSource;
+  private AccessibilityNodeInfo previousEventSource;
+  private int yOffset;
+  private int x1Coordinate;
```
Comment (DaveTryon) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java:18]: It looks like x1Coordiate,y1Coordinate is our start point and that x2Coordinate,y2Coordinate is our end point. That's a bit wordy and is only clear if you look at other code. How would you feel about more descriptive names like xStart, yStart, xEnd, yEnd ?

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    this.previousEventSource.getBoundsInScreen(this.prevRect);
+    this.prevRect.offset(0, this.yOffset);
+
+    this.x1Coordinate = currentRect.centerX();
+    this.y1Coordinate = currentRect.centerY();
+    this.x2Coordinate = prevRect.centerX();
+    this.y2Coordinate = prevRect.centerY();
+  }
+
+  private int setYOffset() {
```
Comment (DaveTryon) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java:69]: `FocusElementLine.setYOffset` is identical to `FocusElementHighlight.setYOffset`. Would it make sense to make this a utility function that is used by both classes?

Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java:69]: @DaveTryon I like that idea.  Would you suggest a `FocusElementUtilities` class to handle that call, that is then passed into the constructor of each the `FocusElementLine` and `FocusElementHighlight` classes?

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  private void drawConnectingLine(
+      int x1Coordinate,
+      int y1Coordinate,
+      int x2Coordinate,
+      int y2Coordinate,
+      Paint paint,
+      Canvas canvas) {
+    canvas.drawLine(x1Coordinate, y1Coordinate, x2Coordinate, y2Coordinate, paint);
```
Comment (DaveTryon) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java:88]: Naive question: The figma shows that the connecting line has a semi-opaque background beneath the dashes. Are we taking that as a literal requirement, and can we construct a `Paint` object that can do that in a single call to `drawLine`, or do we need to think about calling `drawLine` once for the background and a second time for the dashes?

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java (hunk 6 above)
Comment (DaveTryon) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java:69]: @brocktaylor7, the approach you describe makes the most sense when we're changing state within a method. The cleanest approach is usually to allow the injected class to own the state, then the containing class reads the state from the injected class. In *this specific case*, there's no state to manage--we can just pass in a view and get back an offset. We can have a simple static helper class something like this:
```
class OffsetHelper {
  public int getYOffset(View view) {
...
  }
}
```

If the helper method has external dependencies (like the call to `getResources().getDimensionPixelSize()`) that make testing difficult, there may be a benefit to use an injected interface instead of a static class. Ultimately, we want to find the "sweet spot" between too little and too much complexity.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java (hunk 4 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java]: I've removed the TAGs.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java (hunk 3 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementHighlight.java]: This was a good suggestion, I've updated the code to do this.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java (hunk 5 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java]: I like using start and end, it's definitely more succinct and still very clear.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java (hunk 6 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java]: Went with a static helper class. PowerMock allows for mocking the system dependencies as well so we're able to test any non-system functionality

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java (hunk 7 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java]: Dave, this was a great catch, I've adjusted the styles to draw the background as well.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusCanvasView.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import android.view.View;
+import java.util.ArrayList;
+
+public class FocusCanvasView extends View {
+  private ArrayList<FocusElementHighlight> focusElementHighlights;
+  private ArrayList<FocusElementLine> focusElementLines;
+
+  public FocusCanvasView(Context context) {
+    super(context);
+    this.focusElementHighlights = new ArrayList<>();
```
Comment (waabid) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusCanvasView.java:17]: Seems like these are set externally so we should not need to create them here?

Comment (waabid) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusCanvasView.java:17]: I guess if there is a possibility of the draw method being invoked when these are not set, then we should specifically handle that scenario there instead of this.

Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusCanvasView.java]: You're right, I've now removed that and handled the null case on draw. Good catch!

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+    if (this.focusElementHighlights.size() > 1) {
+      this.setPreviousElementHighlightNonCurrent(
+          this.focusElementHighlights.get(this.focusElementHighlights.size() - 2));
+    }
+    if (focusElementLines.size() > 1) {
+      this.setPreviousLineNonCurrent(this.focusElementLines.get(this.focusElementLines.size() - 2));
+    }
+
+    this.focusCanvasView.setFocusElementHighlights(this.focusElementHighlights);
```
Comment (waabid) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java:50]: nit: seems like this can just be one method for setting the focus highlights/lines.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+      FocusCanvasView focusCanvasView) {
+    this.context = context;
+    this.focusElementHighlights = new ArrayList<>();
+    this.focusElementLines = new ArrayList<>();
+    this.tabStopCount = 0;
+    this.styles = focusVisualizerStyles;
+    this.focusCanvasView = focusCanvasView;
+  }
+
+  public void HandleAccessibilityRedrawEvent(AccessibilityEvent event) {}
```
Comment (waabid) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java:31]: Missing call to redraw here?

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+      return null;
+    }
+    return this.focusElementHighlights.get(this.focusElementHighlights.size() - 1).getEventSource();
+  }
+
+  public void resetVisualizations() {
+    this.tabStopCount = 0;
+    this.focusElementHighlights = new ArrayList<>();
+    this.focusElementLines = new ArrayList<>();
+    this.focusCanvasView.redraw();
```
Comment (waabid) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java:97]: Would you not want to use `setFocusElementHighlights` (and the other one) here?

Comment (waabid) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java:97]: Since you're creating new objects that is. Alternatively, I think ArrayList has removeAll.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    for (int i = 0; i < this.focusElementHighlights.size(); i++) {
+      this.focusElementHighlights.get(i).updateWithNewCoordinates();
+    }
+    for (int i = 0; i < this.focusElementLines.size(); i++) {
+      this.focusElementLines.get(i).updateWithNewCoordinates();
+    }
+    this.focusCanvasView.redraw();
+  }
+
+  public void orientationChangedHandler() {
```
Comment (waabid) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java:110]: Orientation changes fire the same events as the events that would cause a redraw. I think we should be updating new coordinates every time we do that.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java (hunk 9 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java]: updated to a single method

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java (hunk 10 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java]: Good catch, I've added in the redraw

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java (hunk 12 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java]: I've moved the handling of this to the HandleAccessibilityRedrawEvent

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java (hunk 11 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java]: used ArrayList.clear() instead, does the same thing, but it's more performant.

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+    AccessibilityNodeInfo eventSource = event.getSource();
+    AccessibilityNodeInfo previousEventSource = this.getPreviousEventSource();
+
+    this.createFocusElementHighlight(eventSource);
+    this.createFocusElementLine(eventSource, previousEventSource);
+
+    if (this.focusElementHighlights.size() > 1) {
+      this.setPreviousElementHighlightNonCurrent(
+          this.focusElementHighlights.get(this.focusElementHighlights.size() - 2));
```
Comment (DaveTryon) [inline AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java:58]: It's not immediately apparent what's special about the value of 2, making it a magic number. Would it help code readability to use a named constant here?. Presumably the same constant would also be used in line 61.

**Review by DaveTryon — APPROVED** [review]
I haven't run this locally but the code looks good! I left one minor suggestion

File: AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java (hunk 13 above)
Comment (brocktaylor7) [reply AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java]: @DaveTryon this is a good suggestion, there's no reason that we can't create the new elements after we set the previous elements as non-current, which would allow us to use -1 instead of -2 to just get the last element in the current list.  I've updated the code to handle it this way!
